File: Python/remove_duplicates.py
# Given a sorted array nums, remove the duplicates in-place such
# that each element appear only once and return the new length.

# Do not allocate extra space for another array, you must do this
# by modifying the input array in-place with O(1) extra memory.

def removeDuplicates(nums):
    j = 1

    for i in range(0, len(nums)):
        try:
            while (nums[j] == nums[i]):
                del nums[j]

        except IndexError:
            break

        j += 1

    return nums

# # Alternative easy peasy solution
# def removeDuplicates(nums):
#     # Create a set out of input, then turn back to list, and return length
#     return len(list(set(nums)))


# print(removeDuplicates([1, 1, 1, 2, 2, 2, 2]))
# print(removeDuplicates([1, 1, 1, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5]))
# print(removeDuplicates([10, 10, 20, 20, 20, 20, 30, 30, 40, 40, 50]))

File: Javascript/removeDuplicates.js
// Given a sorted array nums, remove the duplicates in-place such
// that each element appear only once and return the new length.

// Do not allocate extra space for another array, you must do this
// by modifying the input array in-place with O(1) extra memory.

const removeDuplicates = (nums) => {
  // If array weren't sorted, sort in place
  // nums.sort();
  let j = 1;

  for (let i = 0; i < nums.length; i++) {
    while (nums[j] === nums[i]) {
      // Splice out 1 element
      nums.splice(j, 1);
    }
    j++;
  }
  return nums.length;
};

// Easy peasy version
// const removeDuplicates = (nums) => {
//   return [...new Set(nums)].length;
// };

// console.log(removeDuplicates([10, 10, 10, 20, 30, 30]));
// console.log(removeDuplicates([1, 1, 2, 2, 3, 3, 4, 4]));
// console.log(removeDuplicates([10, 30, 40, 50, 20, 20, 30]));
// console.log(removeDuplicates([5, 5, 5, 5]));
// console.log(removeDuplicates([1, 1, 2]));
